homeassistant/components/blue_current/__init__.py

```diff
@@ -0,0 +1,178 @@
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ATTR_NAME, CONF_API_TOKEN, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.event import async_call_later
+
+from .const import DOMAIN, EVSE_ID, LOGGER, MODEL_TYPE
+
+PLATFORMS = [Platform.SENSOR]
+CHARGE_POINTS = "CHARGE_POINTS"
+DATA = "data"
+SMALL_DELAY = 1
+LARGE_DELAY = 20
+
+GRID = "GRID"
+OBJECT = "object"
+VALUE_TYPES = ["CH_STATUS"]
+
+
+async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Set up Blue Current as a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+    client = Client()
+    api_token = config_entry.data[CONF_API_TOKEN]
+    connector = Connector(hass, config_entry, client)
+
+    try:
+        await connector.connect(api_token)
+    except InvalidApiToken:
+        LOGGER.error("Invalid Api token")
+        return False
+    except BlueCurrentException as err:
+        raise ConfigEntryNotReady from err
+
+    hass.async_create_task(connector.start_loop())
+    await client.get_charge_points()
+
+    await client.wait_for_response()
+    hass.data[DOMAIN][config_entry.entry_id] = connector
+    await hass.config_entries.async_forward_entry_setups(config_entry, PLATFORMS)
+
+    config_entry.async_on_unload(connector.disconnect)
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Unload the Blue Current config entry."""
+
+    unload_ok = await hass.config_entries.async_unload_platforms(
+        config_entry, PLATFORMS
+    )
+    if unload_ok:
+        hass.data[DOMAIN].pop(config_entry.entry_id)
+
+    return unload_ok
+
+
+class Connector:
+    """Define a class that connects to the Blue Current websocket API."""
+
+    def __init__(
+        self, hass: HomeAssistant, config: ConfigEntry, client: Client
+    ) -> None:
+        """Initialize."""
+        self.config: ConfigEntry = config
+        self.hass: HomeAssistant = hass
+        self.client: Client = client
+        self.charge_points: dict[str, dict] = {}
+        self.grid: dict[str, Any] = {}
+        self.available = False
+
+    async def connect(self, token: str) -> None:
+        """Register on_data and connect to the websocket."""
+        await self.client.connect(token)
+        self.available = True
+
+    async def on_data(self, message: dict) -> None:
+        """Handle received data."""
+
+        async def handle_charge_points(data: list) -> None:
+            """Loop over the charge points and get their data."""
+            for entry in data:
+                evse_id = entry[EVSE_ID]
+                model = entry[MODEL_TYPE]
+                name = entry[ATTR_NAME]
+                self.add_charge_point(evse_id, model, name)
+                await self.get_charge_point_data(evse_id)
```

Comment:
Can we gather this work with `asyncio.gather` or does it need to be done sequentially?